### ISSUE TITLE
Scroll jump distance is too large

### ISSUE THREAD

**Issue by vgaming** [wesnoth/wesnoth#3018]
Scroll jump distance is too large. [My] eye doesn't recognize such jump as a "scroll". It feels to be a total re-paint instead, making visual navigation game list hard. "Very hard" I'd even say, though this is subjective.

Screenshot before a single scroll event:
![scroll-jump-1](https://user-images.githubusercontent.com/33452702/39652081-071660c6-4ff5-11e8-8309-f9d928f974a2.png)
After:
![scroll-jump-2](https://user-images.githubusercontent.com/33452702/39652087-10422e5a-4ff5-11e8-9b6b-447561b25a54.png)

The lower your monitor resolution the worse it'll feel.

I've compared it with my browser, Firefox. Firefox scroll distance is about 6 times shorter, in pixels.

**Comment by vgaming** [comment]
This is subjective. I ask to comment and agree or disagree if you scrolled the list and get your personal estimation.

**Comment by ProditorMagnus** [comment]
Only used arrows to scroll gamelist so far, but chat log scroll step feels too large.

**Comment by Wedge009** [comment]
I'm inclined to agree it's a bit on the large side.

**Comment by shikadiqueen** [comment]
From a cursory glance at the code, it seems like the scrolling distance is actually larger the more items there are in a container, rather than being constant.

**Comment by Wedge009** [comment]
Proportional, then? Should we make it constant?

**Comment by shikadiqueen** [comment]
I haven't paid enough attention to other major toolkits like Windows, Gtk 3 and Qt 5, but I'd suspect that they generally use constant scrolling steps. To me it feels like that would make the most sense per the principle of least surprise.

For reference, I landed on that method after looking for the scrollwheel handlers in `gui2::scrollbar_container`. They both call `gui2::scrollbar_base::scroll()` with `scrollbar_base::HALF_JUMP_BACKWARDS/HALF_JUMP_FORWARD`, and you can see the formula used in the method's implementation — although I didn't bother with tracking down every single one of the variables involved, it's very obvious that the new position is directly proportional to the `visible_items_` field, and it does feel that way in practice, especially after experimenting with a factor of 0.25 instead of 0.5 and finding completely different results with a long list such as the unit list in the Create Unit dialog as opposed to shorter ones like the Campaigns menu.

**Comment by shikadiqueen** [comment]
For reference, both Plasma and Windows have settings that imply that scrolling is by a constant number of lines, which probably means it's proportional to the average size of each row, but not to the total row count.

**Plasma:**

![image](https://user-images.githubusercontent.com/489895/41768302-ce44ab4e-75d9-11e8-94a3-26f27ff8ccdd.png)

**Windows:**

![image](https://user-images.githubusercontent.com/489895/41768419-292566a2-75da-11e8-8f78-bfc6bbb683ba.png)

**Comment by gfgtdf** [comment]
> For reference, both Plasma and Windows have settings that imply that scrolling is by a constant number of lines, which probably means it's proportional to the average size of each row


Ye i'd be curious about what exactly i means are in a 'listbox' case liek here, woul 'lines' actually mean 'listbox items' or 'text lines ' (which would clearly be less than a listbox items height)

`visible_items_ ` is the numebr of shown items in the curent shows rectange, in most cases this is basicially the height of the widget in pixels, and `item_count_` is the total size of the `'inner' list (in pixels)


I think its quite possible that the current behviour is there because of technical reasons it could for example be the case that: scollbar_base does not know whether 'items' are actuall the rows of a listbox or just the number of pixels in a 'unstructured scroll area' (i actull think it's always the number of pixels nowdays, again not sure), so the code does not want to use scrolling based in it beasue it might result in scrolling muhc faster on listboxes than on

**Comment by shikadiqueen** [comment]
> I think its quite possible that the current behviour is there because of technical reasons

Yeah, my assumption as well is that it's left to the users of `scrollbar_base` to decide what the number of "items" stands for. For example, a scrolling pane has only one "item" (a grid) as far as a user is concerned), but it needs to be scrolled all the same.

Like I said, for a listbox it would make sense to make the scrolling step a constant number of rows (for a vertical listbox) or columns (for a horizontal one). The exact size of said step obviously depends on the size of said elements, so presumably the listbox should calculate it and set it on the scrollbars based on the average height/width of the elements.

**Comment by Wedge009** [comment]
As an experiment against 1.14.9+dev I switched `HALF_JUMP...` for `ITEM_...` in `scrollbar_container.cpp`:
https://github.com/wesnoth/wesnoth/blob/7c76f8d0665068f6b5544caa406b9303907d769f/src/gui/widgets/scrollbar_container.cpp#L1138-L1162

I hoped that the scroll would then switch to one MP game at a time. Unfortunately, instead, it seemed to scroll only one _pixel height_ per scroll wheel notch, which is infeasibly slow. I think we might need something in between the single `ITEM` and the variable `JUMP` scrolling.
